### ISSUE TITLE
Syntax of storage and memory handling

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/fe#57]
### Storage and memory

This isn't actionable but it's a thought I wanted to get written down to pick back up when the time comes.
Solidity has two keywords `storage` and `memory` that can control where a value ends up being stored.

Example:

```solidity
    function example1(uint storage value) public {
        // code
    }
    function example2(uint memory value) public {
        //code
    }
```
The Rust language has a philosophy that often prefers to make things *library features*  rather than *syntax features*  if possible. (That isn't to say rust has a simple syntax, it sure has a lot of syntax rules).

For instance, in Rust, to store a `u256` on the heap rather than the stack one would but it in a `Box<T>` so the type would end up being `Box<u256>` and no special keyword is needed. I find that to be superior in many circumstances. Not only does it mean less keywords, it also means that the language can evolve easier without introducing breaking changes to its syntax. For instance, at some point a `BetterBox<T>` may be introduced and one could even end up using both side by side. Alternatively versions of `Box<T>` could live behind different import paths.

So, to translate the Solidity example into a made up form of future Fe code, this is what I could see it end up looking:

```
    pub def example1(value: Storage<u256>):
        # code
    pub def example2(value: Memory<u256>):
        # code
```

**Comment by g-r-a-n-t** [comment]
If we decide to add a feature like this, I think this is the way to go. I do wonder, though, if it's possible to completely avoid the need for this.

Say we have this function:

```
def foo(bar: u256[100]) -> u256:
    return bar[42]
```

Currently, this would be mapped to something like:

```
function foo(bar) -> val {
    val := _mem_array_index(bar, 42) // note: there is no runtime function `_mem_array_index` atm, but we should have one
}
```

I'm thinking we could create multiple versions of the same Fe function in Yul, but with different parameter locations to accommodate all invocations.

For example, let's say `foo` is called two different ways within the contract:
1. `self.foo(my_mem_array)`
2. `self.foo(self.my_sto_array)`

The semantic analysis pass would notice this and determine that two separate `foo` functions need to be created. One where `bar` is in memory and the other where `bar` is in storage i.e.:

```
function foo_mem(bar) -> val {
    val := _mem_array_index(bar, 42)
}

function foo_sto(bar) -> val {
    val := _sto_array_index(bar, 42)
}
```

Then the two function calls would be mapped like so:
1. `self.foo(my_mem_array)` -> `foo_1(my_mem_array)`
2. `self.foo(self.my_sto_array)` -> `foo_2(<storage index>)`

**Comment by cburgdorf** [comment]
I agree that this could be inferred for the parameters to the function but I assume there are still situations (e.g. contract member variables) where the programmer needs/wants to be in control whether something is stored in memory or on the storage.

**Comment by davesque** [comment]
Hey, just noticed recently that you guys are continuing work on the lang.  Awesome stuff!

Thought I'd jump in and comment that you might want to consider this syntax extension against the already existing syntax for declaring indexed event fields and public contract fields and methods.  In other words, would it make sense to also convert indexed field declarations to looking like this?
```python
# old
event Signed:
    idx book_msg: BookMsg

# new
event Signed:
    book_msg: Indexed<BookMsg>
```
I think maybe not for some of the reasons I discuss in this thread: https://github.com/ethereum/fe/issues/8.  Namely, it might get cumbersome to stack multiple annotation on anything that uses any kind of "calling" syntax where pairs of delimiters need to enclose some argument.  So in an imaginary future where we want to annotate a single variable with multiple properties, it might start to look a bit hairy:
```python
# with angle bracket call-like syntax
def myFunc(foo: SomeAnnotation<Storage<u256[100]>>) -> u256:
    ...

# with prefix qualifier syntax
def myFunc(som_annotation storage foo: u256[100]) -> u256:
    ...

# with prefix qualifier syntax
def myFunc(som_annotation storage foo: u256[100]) -> u256:
    ...
```
The latter might be a bit more readable...maybe.  Actually, I don't know how I feel about it now that I'm seeing it typed out.  But I guess the point is that there seems to be some overlap between these different syntaxes and it would be good to think about if that will ever seem confusing or if they should be consolidated somehow.

**Comment by g-r-a-n-t** [comment]
One argument that could be made on behalf of using angle brackets here is that the programmer could shorten things up using type definitions.

```
type Custom<T> = SomeAnnotation<Storage<T>>

# with angle bracket call-like syntax
def myFunc(foo: Custom<u256[100]>) -> u256:
    ...
```

My hope, though, is that the situations requiring compound annotations would be limited, as they take a toll on the languages accessibility regardless of the syntax.

**Comment by cburgdorf** [comment]
I think that if we go down that route of turning these things that are currently kewords (e.g. `indexed`) into a generic builtin like `Indexed<T>` then it would be crucial to go along with a std library and imports.

E.g. 

```
from std import Indexed, Storage, Memory

# Now usage as Indexed<uint256>, Storage<address>, Memory<string> etc
```

Because, if we don't do that, then it seems to me that we would be giving up on the mentioned benefits of library features vs syntax features (e.g. improved types importable from a different path).

I've also recently read that a std library is one of the things that the solidity team has on their radar and @pipermerriam also mentioned it as one of the things being on his personal wish list. I think this would naturally go hand in hand.

**Comment by davesque** [comment]
Grant:
> One argument that could be made on behalf of using angle brackets here is that the programmer could shorten things up using type definitions.

Christoph:
> I think that if we go down that route of turning these things that are currently kewords (e.g. `indexed`) into a generic builtin like `Indexed<T>` then it would be crucial to go along with a std library and imports.
> ...
> I've also recently read that a std library is one of the things that the solidity team has on their radar and @pipermerriam also mentioned it as one of the things being on his personal wish list. I think this would naturally go hand in hand.

Both great points, guys.  In that case, I feel better about favoring the generics syntax.  And as I mentioned already, I didn't really like the way those prefixes looked when they appeared together in my previous example.

I think there's still a question of how these generic-like annotations line up conceptually with prefix annotations such as `pub` on methods and fields.  Previously, I had sort of conceptually lumped the `idx` prefix together with the `pub` prefix.  Now, we're looking at breaking those things up into different syntaxes.  So I wonder, is there anything that should happen to `pub`?  Should fields be annotated with a `Public<T>` type?  Should we also break down and just use decorator-like syntax for method annotations (`@public`) and get rid of the `pub` prefix?

**Comment by g-r-a-n-t** [comment]
> then it would be crucial to go along with a std library and imports

We could probably just import the std library implicitly. This way we don't need to worry about implementing imports now just for the sake of the std library. It seems like most languages do this anyway.

> So I wonder, is there anything that should happen to pub? Should fields be annotated with a Public<T> type?

I think `pub` is good where it's at. If we're using Rust as a model language for library features and where you should draw the line between library and syntax features, then leaving public/private as a syntax feature is the correct way to go. And if we are going to leave it as a syntax feature, then I think the `pub` prefix looks best.

**Comment by cburgdorf** [comment]
>We could probably just import the std library implicitly.

Yes, I thought about that, too and I think it is ok to begin with. I do think that - at a later stage - it is valuable to opt-out of it and conditionally import (e.g. to decide between different versions of built-ins)

>I think pub is good where it's at. If we're using Rust as a model language for library features and where you should draw the line between library and syntax features, then leaving public/private as a syntax feature is the correct way to go. And if we are going to leave it as a syntax feature, then I think the pub prefix looks best.

Yes, I'm on the same page with `pub`. I want to add that I just recently [opened up the same discussion about using decorators for visibility](https://github.com/ethereum/fe/pull/51#issuecomment-698388800) and I would stay open to the thought if it turns out that we will have to support a range of different modifiers (e.g. `public`, `external`, `internal` whatever)
For now, I think going forward with `pub` is the right approach.

**Comment by davesque** [comment]
Alright, cool.  It sounds like the proper conceptual difference is this lib vs syntax feature distinction.  I think it sounds good.
